src/core/SampleBuffer.cpp

```diff
@@ -96,621 +70,271 @@ SampleBuffer::SampleBuffer(const QString & audioFile, bool isBase64Data)
 	}
 	else
 	{
-		m_audioFile = audioFile;
-		update();
+		loadFromAudioFile(audioFile);
 	}
 }
 
-
-
-
 SampleBuffer::SampleBuffer(const sampleFrame * data, const f_cnt_t frames)
 	: SampleBuffer()
 {
 	if (frames > 0)
 	{
-		m_origData = MM_ALLOC<sampleFrame>( frames);
-		memcpy(m_origData, data, frames * BYTES_PER_FRAME);
-		m_origFrames = frames;
+		m_data = std::vector<sampleFrame>(data, data + frames);
+		setAllPointFrames(0, frames, 0, frames);
 		update();
 	}
 }
 
-
-
-
 SampleBuffer::SampleBuffer(const f_cnt_t frames)
 	: SampleBuffer()
 {
 	if (frames > 0)
 	{
-		m_origData = MM_ALLOC<sampleFrame>( frames);
-		memset(m_origData, 0, frames * BYTES_PER_FRAME);
-		m_origFrames = frames;
+		m_data = std::vector<sampleFrame>(frames);
+		setAllPointFrames(0, frames, 0, frames);
 		update();
 	}
 }
 
-
-
-
 SampleBuffer::SampleBuffer(const SampleBuffer& orig)
 {
-	orig.m_varLock.lockForRead();
-
+	const auto lockGuard = std::shared_lock{orig.m_mutex};
 	m_audioFile = orig.m_audioFile;
-	m_origFrames = orig.m_origFrames;
-	m_origData = (m_origFrames > 0) ? MM_ALLOC<sampleFrame>( m_origFrames) : nullptr;
-	m_frames = orig.m_frames;
-	m_data = (m_frames > 0) ? MM_ALLOC<sampleFrame>( m_frames) : nullptr;
-	m_startFrame = orig.m_startFrame;
-	m_endFrame = orig.m_endFrame;
-	m_loopStartFrame = orig.m_loopStartFrame;
-	m_loopEndFrame = orig.m_loopEndFrame;
+	m_data = orig.m_data;
+	m_playMarkers = orig.m_playMarkers;
 	m_amplification = orig.m_amplification;
 	m_reversed = orig.m_reversed;
 	m_frequency = orig.m_frequency;
 	m_sampleRate = orig.m_sampleRate;
-
-	//Deep copy m_origData and m_data from original
-	const auto origFrameBytes = m_origFrames * BYTES_PER_FRAME;
-	const auto frameBytes = m_frames * BYTES_PER_FRAME;
-	if (orig.m_origData != nullptr && origFrameBytes > 0)
-		{ memcpy(m_origData, orig.m_origData, origFrameBytes); }
-	if (orig.m_data != nullptr && frameBytes > 0)
-		{ memcpy(m_data, orig.m_data, frameBytes); }
-
-	orig.m_varLock.unlock();
 }
 
-
-
-
 void swap(SampleBuffer& first, SampleBuffer& second) noexcept
 {
 	using std::swap;
 
-	// Lock both buffers for writing, with address as lock ordering
 	if (&first == &second) { return; }
-	else if (&first > &second)
-	{
-		first.m_varLock.lockForWrite();
-		second.m_varLock.lockForWrite();
-	}
-	else
-	{
-		second.m_varLock.lockForWrite();
-		first.m_varLock.lockForWrite();
-	}
+	const auto firstLockGuard = std::unique_lock{first.m_mutex};
+	const auto secondLockGuard = std::unique_lock{second.m_mutex};
```

Comment:
Given that this is happening in a class derived from ``QObject``, would having a copy constructor or copy assignment even be useful, since ``QObject``'s cannot have copy constructors/assignment operators?